en/middleware/mavlink.md

```diff
@@ -2,200 +2,419 @@
+		// Loop through _battery_status_subs (subscription to array of BatteryStatus instances)
+		for (auto &battery_sub : _battery_status_subs) {
+            // battery_status_s is a struct that can hold the battery object topic
+			battery_status_s battery_status;
 
-            _msg_ca_trajectory.timestamp = _ca_trajectory.timestamp;
-            _msg_ca_trajectory.time_start_usec = _ca_trajectory.time_start_usec;
-            _msg_ca_trajectory.time_stop_usec  = _ca_trajectory.time_stop_usec;
-            _msg_ca_trajectory.coefficients =_ca_trajectory.coefficients;
-            _msg_ca_trajectory.seq_id = _ca_trajectory.seq_id;
+			// Update battery_status and publish only if the status has changed
+			if (battery_sub.update(&battery_status)) {
+                // mavlink_battery_status_demo_t is the MAVLink message object
+				mavlink_battery_status_demo_t bat_msg{};
 
-            mavlink_msg_ca_trajectory_send_struct(_mavlink->get_channel(), &_msg_ca_trajectory);
-            
-            return true;
-        }
+				bat_msg.id = battery_status.id - 1;
+				bat_msg.battery_remaining = (battery_status.connected) ? roundf(battery_status.remaining * 100.f) : -1;
+
+				// check if temperature valid
+				if (battery_status.connected && PX4_ISFINITE(battery_status.temperature)) {
+					bat_msg.temperature = battery_status.temperature * 100.f;
+				} else {
+					bat_msg.temperature = INT16_MAX;
+				}
+
+                //Send the message
+				mavlink_msg_battery_status_demo_send_struct(_mavlink->get_channel(), &bat_msg);
+				updated = true;
+			}
+		}
+
+		return updated;
+	}
 
-        return false;
-    }
 };
 ```
 
+Most streaming classes are very similar (see examples in [/src/modules/mavlink/streams](https://github.com/PX4/PX4-Autopilot/tree/main/src/modules/mavlink/streams)):
+
+- The streaming class derives from [`MavlinkStream`](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_stream.h) and is named using the pattern `MavlinkStream``<CamelCaseMessageName>`.
+- The `public` definitions are "near-boilerplate", allowing PX4 to get an instance of the class (`new_instance()`), and then to use it to fetch the name, id, and size of the message from the MAVLink headers (`get_name()`, `get_name_static()`, `get_id_static()`, `get_id()`, `get_size()`).
+  For your own streaming classes these can just be copied and modified to match the values for your MAVLink message.
+- The `private` definitions subscribe to the uORB topics that need to be published.
+  Here we also define constructors to prevent the definition being copied.
+- The `protected` section is where the important work takes place!
+  Here we override the `send()` method, copying values from the subscribed uORB topic(s) into appropriate fields in the MAVLink message, and then send the message.
+
+  In this case the uORB topic has multiple instances: one for each battery.
```

Comment:
Downside of battery is that it support multiple instances so more complicated processing. Upside is that we get to discuss. I wasn't aware for example difference between subscription `update()` and `copy()`. Section below needs to be checked carefully.